torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
                     )
             eod.param_order.append(param_index)
 
+    def register_comm_hook(self, state: object, hook: callable):
+        r"""
+
+        Registers a communication hook which is an enhancement that provides a
+        flexible hook to users where they can specify how FSDP aggregates gradients
+        across multiple workers.
+
+        This hook can be used to implement several algorithms like GossipGrad
+        and gradient compression which involve different communication strategies for
+        parameter syncs while running Fully Sharded Data Parallel training.
+
+        .. warning::
+            FSDP only support communication hooks for a NO_SHARD strategy at this time.
+            If other strategies are used, an error will be raised.
+
+        Args:
+            state (object): Passed to the hook to maintain any state information during the training process.
+                            Examples include error feedback in gradient compression,
+                            peers to communicate with next in GossipGrad, etc.
+
+                            It is locally stored by each worker
+                            and shared by all the gradient tensors on the worker.
+
+            hook (callable): Callable with the following signature:
+                             ``hook: Callable[torch.Tensor] -> None``:
+
+                             This function takes in a python tensor, which represents a gradient
+                             that needs to be communicated across ranks. It then performs
+                             all neccessary processing and returns nothing.
+
+        """
+        if self.sharding_strategy != ShardingStrategy.NO_SHARD:
+
+            raise RuntimeError(
+                "Communication hooks are currently only available for a NO_SHARD strategy."
+            )
+        else:
+            # register same hook for root and all submodules
+            for submodule in self.fsdp_modules(self):
+                if submodule.check_is_root():
+                    logger.warn(
```

Comment:
I would advocate for removing this warning and just inserting it into the documentation, it is expected behavior so not really worth raising a warning imo

Reply:
Removed warning from a logger